Rework explainer (#47)
WICG/badging: explainer.md

  app](https://www.w3.org/TR/appmanifest/#installable-web-applications) icon.

Here is a mock of a badge being applied to the page tab:

![Mock of a badge being applied to the page tab](images/tab-badge.png)

This use case is already satisfied today with sites dynamically setting either
their favicon or title to include some status indicator. The use of an explicit
badge API has a number of advantages over the "hack" ways:

* The badge can appear outside (completely, or partially overlapping) of the
  favicon, leaving more room for the site's brand image.
* The badge is meaningful to the user agent, which can present it to the user in
  various ways, such as announcing it verbally to a user using a screen reader.
* Badges can be displayed with a consistent style across websites, chosen by the
  user agent.
* User agents can provide a way for users to disable badges on a per-site or
  global basis.

The "handle" context is more nebulous because it means associating a badge with
any place the user agent shows an icon representing a site, page or app. For
example, it could be applied to an icon in the bookmark bar, or a "commonly
visited sites" index on the new tab page.

> ![Mock of a badge being applied to the bookmark bar](images/bookmark-badge.png)

Reviewer: The image is really small, so you don't really get any context of what the picture is of. Maybe use a more zoomed out image?

Author: Done.

![Android home screen badge](images/android-badge.png)
<br>Android home screen badge

Unlike document contexts, badges applied in "handle" contexts can be shown and
updated even when there are no tabs or windows open for the app. So there are
some special considerations for this use case, such as how to set the badge in
response to a [Push message](https://www.w3.org/TR/push-api/), and how to scope
a badge so that it appears on the desired set of pages or apps. The most
commonly requested use case for "handle" badging is badging an app icon on the
OS shelf, but it generalizes to non-installed sites as well.

## Usage examples

The simplest possible usage of the API is a single call to `Badge.set` from a
foreground context (which might be used to show an unread count in an email
app):

```js
Badge.set(getUnreadCount());
```

This will set the badge for all pages and apps in the current origin until it is

> changed. If `getUnreadCount()` (the argument to `Badge.set`) is 0, it will

Author: Added that to the "Model" section. I said this:

> The user agent is allowed to clear the badge whenever there are no foreground pages open on the origin (the intention of this is so that when the user agent quits, it does not need to serialize all the badge data and restore it on start-up; sites should re-apply the badge when they open).

Is that OK? It's a bit hard to exactly define when UAs are allowed to clear badges.

> changed. If `getUnreadCount()` (the argument to `Badge.set`) is 0, it will

Reviewer: :+1:

automatically clear the badge.

If you just want to show a status indicator flag without a number, use the
Boolean mode of the API by calling `Badge.set` without an argument, and
`Badge.clear` (which might be done to indicate that it is the player's turn to
move in a multiplayer game):

```js
if (myTurn())
  Badge.set();
else
  Badge.clear();
```

The reason we are considering both the "document" and "handle" contexts in the
same API, and not two separate badging APIs, is so that in the common case,
developers can just set a badge for the origin and have it show up in whatever
places the user agent wants to show it. However, if you just want to badge a
specific set of URLs, and not the whole site, use the `scope` option (which
might be done if you have a different "unread count" on each page):

```js

> Badge.set(getUnreadCount(location.pathname), {scope: location});

Author: Mm, getPageNotifications() is less clear to me that it's getting a count specific to this current page.

```

The scope is a URL prefix; the badge is applied on all pages whose URL [starts

> with](https://www.w3.org/TR/appmanifest/#dfn-within-scope) that prefix.

Author: Maybe... it doesn't really matter. I prefer the app definition (because I wrote it -- ha!) but mostly because it's got notes that explain why it does a prefix match instead of the obvious one.

> with](https://www.w3.org/TR/appmanifest/#dfn-within-scope) that prefix.

Reviewer: No strong feelings here.

More advanced examples are given in a [separate document](docs/examples.md).

## Goals and use cases

The purpose of this API is:

* To subtly notify the user that there is new activity that might require their attention without requiring an OS-level [notification](https://notifications.spec.whatwg.org/).
* To indicate a small amount of additional information, such as an unread count.

> * To allow certain user-blessed pages (such as Bookmarks or [Installed Web Applications](https://www.w3.org/TR/appmanifest/#installable-web-applications)) to convey this information, regardless of whether they are currently open.

Author: Done.

Examples of sites that may use this API:

* Chat, email, and social apps could signal that new messages have arrived.
* Any application that needs to signal that user action is required (e.g., in a turn-based game, when it is the player's turn).
* As a permanent indicator of a page's status (e.g., on a build page, to show that the build has completed).

### Use with Notifications

The Badge API goes hand-in-hand with Notifications, since both APIs provide a
way to give the user updates when they aren't directly looking at the page.
Arguably, Badge could be part of Notifications, but we view it as a separate,
complementary API.

Advantages of using the badging API over notifications:

* Can be used for much higher frequency / lower priority events than notifications, because each new event does not disrupt the user.
* There may be no need to request permission to use the badging API, since it is much less invasive than a notification.

Typically, sites will want to use both APIs together: notifications for high-importance events such as new direct messages or incoming calls, and badges for all new messages including group chats not directly addressed to the user.

> 

Reviewer: Do we want some kind of comment to this end still? 

In the short term, we aren't going to support badging from a push event, but you'll still be able to use the API from the ServiceWorker (background sync, fetch, whatever).

Author: Yes, there's a lot about this in the Push section. I think it used to be misleading that we said not to use from a SW, when we really meant that you couldn't use it from a push without also showing a notification. Let's not tie those two restrictions together.

## Badge scope

This section explains the difference between badging a *set of URLs* versus
badging a *document*.

Consider four increasingly narrow sets of pages that a Badge is applied to:

1. An entire origin (e.g., when you just want to set an unread count for the
   entire site).
2. All pages whose URL [starts

>    with](https://www.w3.org/TR/appmanifest/#dfn-within-scope) a given path

Reviewer: ```suggestion
   with](https://w3c.github.io/ServiceWorker/#service-worker-registration-scope) a given path

   (e.g., if a site is organised into "projects", each with their own status
   indicator).

> 3. All pages with a specific URL path (e.g., if each page has a separate status

Reviewer: I was thinking about this. I don't think scope is quite the solution we think it is.

It won't work for 
- Query strings (already discussed)
- SPAs on a fragment url (similar to query strings)
- Apps that serve pages on `/`, `/app`, `/app/page` and `/app/page/1` (there's no way to badge `/app` and only `/app`).

Author: I don't think any of these really change our calculation here. I think it's quite unlikely that people are going to want to badge arbitrary groupings of pages. That's why it's broken into these four categories. Almost always, you'll want to badge an entire site/app (which may be at a sub-path if your app shares an origin with other apps, but will _always_ have its own path).

Sometimes you might want to badge individual pages, which will work fine if you designed your URLs correctly. If not, you'll have to use the dynamic badge API.

This solves all the problems I care about.

Reviewer: I kind of wish we didn't have to care about app scope either. Perhaps we could get away with not only supporting origin badges (if you have multiple apps on the same origin, tough, use the dynamic badging API).

It'd solve 90% of the use cases and be waaaay simpler. 

But probably not that big a deal.

   indicator, but all tabs open at a certain URL should share status).
4. A specific document (e.g., if many tabs can be open with URLs on the same
   path, but might have different status, due to either a different query
   string, or different dynamic context inside the document).

Number 3 and 4 seem very similar, but architecturally are very different. In
3, the badge is associated with a URL (and any current or future documents open
at that URL will show that badge), whereas in 4, the badge is associated with a
particular document — other tabs open to the same URL will not share the badge.

What this boils down to is two different modes:

* Badging a particular URL scope, applied to the set of URLs that start with the

>   same path. This covers 1, 2 and 3 above.

Reviewer: Just *barely*, if you *squint* `3`.

Author: Why? Do you mean because of the prefix match thing? Hmm.

I suppose the real problem here is that some "endpoint" paths will not end in a slash, for example "/users/matt", so if you want a per-user badge, you'll have to badge "/users/matt" and NOT "/users/matt/". But that means you're also badging "/users/matthew" and anything else that starts with that prefix.

.... more complexity.

Reviewer: I'm more about the issue where you display a list of users at `/users/` and badge that with a badge which should **only** apply to that page. It'd be kind of weird if it also showed up for `/users/matt`.

Another example:

Builds at `/build/<number>` (check or cross for success/fail)
Build logs at `/builds/<number>/logs` (plain text, shouldn't really have a badge).

Author: Right. So we don't have a way to badge a single URL. But I guess this all goes away if we split the API. The scoped badging is for broad groups of URLs (i.e., an entire app) rather than individual pages.

Reviewer: :+1:

* Badging a particular document, applied to only the browser tab containing the
  document.

The API for requesting to badge the current document, as opposed to a URL set,
is still under consideration. It may simply be a separate option member, e.g.:

```js

> Badge.set(getUnreadCount(), {scopeDocument: true});

Reviewer: Feature detection would still be hard here. Do we raise an exception if `scopeDocument` isn't supported? 

> Probably unimportant note: What about `scopeQuery` and `scopeFragment` options?

Author: Well we either need to separate out the two methods, or add an explicit feature detection as noted in the "Feature detection" section of this doc.

```

When displaying the badge for an app, the user agent should use the badge
matching the app's [scope URL](https://www.w3.org/TR/appmanifest/#scope-member).

### Nested scopes

Since we allow badges to be scoped to different, potentially nested, URLs, it
means that a particular page can be subject to more than one badge at a time.
In this case, the user agent should display only the badge with the most

> specific URL.

Reviewer: ```suggestion
specific [scope](https://w3c.github.io/ServiceWorker/#service-worker-registration-scope).
```

Therefore, clearing a badge (either by calling `Badge.clear()` or
`Badge.set(0)`) does *not necessarily* mean that no badge will be displayed; by

> erasing a badge at one level, a page may inherit a badge from a higher level.

Reviewer: Are we going to propose an API for clearing a badge at a specific scope? It seems kind of important (but also, this is already complicated enough...)

Author: What do you mean? An API for ensuring that no badge is set anywhere above a specific scope?

I wasn't planning on it. I think the recommended practice for this API will be to either:

a) Not nest your badges. Decide on a certain level that badges will exist at, and always set or clear them at at that level, or
b) Be okay with the semantics of nested badges. If you have a site-wide badge and a per-page badge, you must be OK with the per-pages reverting back to the site-wide badge if you clear the page badge. If you want fine-grained control over which badge appears, use the dynamic badge API which doesn't have scope.

In fact, if we separate these out, then really the nested scoping doesn't come up because you won't be using the static badge API for specific pages. You would just use the dynamic scoping.

Reviewer: If we can get away with "What you're doing isn't normal, do the per document thing" I feel like we should sound out doing away with scopes altogether. They're kind of complex.

Author: Well, we still need a way to let you badge individual apps on the same origin. That's basically the only remaining use case for scopes after we split the API, so if we can get the same effect with a simplification, we should, but at the moment, scopes seems the most simple.

For example, consider a site that has made the following two calls:

* `Badge.set(6, {scope: '/users/'});`
* `Badge.set(2, {scope: '/users/1'});`

Now all pages in '/users/' show the badge "6", except for `/users/1`, which
shows the badge "2".

Now if we see `Badge.clear({scope: '/users/1'})`, the pages under `/users/1`
will start showing the badge "6" since that badge is still in effect. If instead
we see `Badge.clear({scope: '/users/'})`, the pages under `/users/1` will still

> show the badge "2", *even if `clear` is called from one of those pages*.

Reviewer: Possibly link to the relevant example /docs/examples.md#badging-for-multiple-apps-on-the-same-origin-as-in-the-case-of-multiple-github-pages-pwas?

Author: Done.

no pages are open.

In this section, we explore two APIs that could be useful for this: [Push
API](https://www.w3.org/TR/push-api/) and [Periodic Background
Sync](https://github.com/WICG/BackgroundSync/blob/master/explainer.md#periodic-synchronization-in-design).

### The Push problem

The [Push API](https://www.w3.org/TR/push-api/) allows servers to send messages
to service workers, which can run JavaScript code even when no foreground page
is running. Thus, a server push could trigger a `Badge.set`.

However, there is a [de facto standard
requirement](https://github.com/w3c/push-api/issues/313) that whenever a push is
received, a notification needs to be displayed. This means it's impossible to
subtly update a badge in the background while no pages are open, without also
displaying a notification.

This may be fine for some use cases (if you want to always show a notification
when updating the badge, possibly with an exception of not showing a
notification if the app is in the foreground). But we specifically designed this
API as a *subtle* notice mechanism that does not require a more distracting
notification in order to set a badge. Because of this de facto requirement, the

> Push API currently isn't suitable for this use case.

Reviewer: At least, not without **significant** modifications.... This makes me so sad 😭

There is also a requirement that the user grants notification permission to
subscribe to push messages.

### Periodic Background Sync

[Periodic Background
Sync](https://github.com/WICG/BackgroundSync/blob/master/explainer.md#periodic-synchronization-in-design)
is a proposed extension to the [Background
Sync](https://wicg.github.io/BackgroundSync/spec/) API, that allows a service
worker to periodically poll the server, which could be used to get an updated
status and call `Badge.set`. However, this API is unreliable: the period that it
gets called is at the discretion of the user agent and can be subject to things

> like battery status.

Reviewer: I don't think the battery status limitation is really a problem, I'd be fine with my badges updating less frequently when I'm low on power. 

Its more that this really isn't the use case the API was designed for (download new articles an hour before the user normally opens the app or something). From a user agent's perspective, it'd be entirely reasonable to limit updates to once per `day`/`hour`

That means when the page isn't open, you could have the badge indicator update
every once in awhile, but have no guarantee that it would be up to date.

(Also, the Periodic Background Sync API is not yet implemented.)

It's possible that the use of Push API and Periodic Background Sync together is
"good enough": high-priority notices (that show a notification) come in through
a Push and are displayed immediately with a notification; low-priority notices
are displayed as badges immediately if there are pages open, or "eventually" if
there are no pages open.

> However, in our view, there is a need for *immediate* update to low-priority

Reviewer: 👍

notices: this is akin to the [urgency vs importance](https://en.wikipedia.org/wiki/Time_management#The_Eisenhower_Method)
dichotomy: a badge is for non-important information but that doesn't mean the
user doesn't want to see the notice in a timely fashion.

### Possible changes to the Push API

Here we discuss ways to potentially allow the Push API to update badges without
showing notifications.

> **Note: This should not be considered blocking.** The Badge API is perfectly

Reviewer: Do we have a preference for

> Note: The foo is barred.

vs 

**Note:** The foo is barred

I have no strong feelings but on the off chance you care

```suggestion
> Note: **This should not be considered blocking.** The Badge API is perfectly
```

Author: I have a (weak) preference for not using quotation blocks for notes.

usable, including from service workers, without these changes, so we consider
these as add-ons that we could introduce at a later time.

#### Waiving the notification requirement

The purpose of the [de facto
requirement](https://github.com/w3c/push-api/issues/313) to show a notification
upon receipt of a push message is to prevent malicious sites from using
high-frequency pushes to do invisible background work, such as crypto mining
(use of the user's computing resources) or monitoring the user's IP address,
giving coarse-grained geographic location (privacy concern). The theory is that
if the site is required to show a notification on each push, the user is at
least going to be aware that the site is constantly doing something, and if it's
unreasonably spammy, the user is likely to revoke the notification permission,
thus erasing the push subscription.

We could make it so that use of the `Badge` API serves the same function as
showing a notification (fulfilling the requirement to use the Push API). But
doing so would probably nullify the reason for this requirement in the first
place: a malicious site that wants to do crypto mining could just set a badge

> every 30 seconds and the user would probably not notice it.

Reviewer: They could set it to the same value, and the user **definitely** wouldn't notice it.

or

```js
const onNotificationReceived = () => {
    Badge.set();
    Badge.clear();
};
```

Author: Done.

high bar for waiving this requirement, e.g., "the user must have [installed the
site](https://www.w3.org/TR/appmanifest/#installable-web-applications) as an
application, and enabled notifications". It may be an acceptable trade-off to
allow these "semi-trusted" sites to perform silent background tasks, in order to
let them keep badges up to date in a timely fashion. This would have to be a
discussion around privacy trade-offs which we (the Badge API authors) aren't
equipped to answer by ourselves.

Technically, the way we would do this is by allowing certain sites to set
`userVisibleOnly` to `false` in the push subscription (which currently has no
defined meaning, and in Chrome at least, is not allowed to be `false`, so there
is currently a de facto requirement that it be set to `true`). If a site is
allowed to set `userVisibleOnly` to `false`, then it can receive badge-only push
messages. If not, it is bound by the existing rules, and must either show a
notification, or turn off low-priority messages.

#### A separate channel for Badge payloads

A more comprehensive solution includes some significant additions to the Push

> API (which the authors of the Badge spec are not really equipped to do).

Reviewer: For my own amusement:

```suggestion
API (which the authors of the Badge spec are really not equipped to do).
```

Author: Done.

> (This has been briefly discussed with Peter Beverloo, an editor of the Push API

Reviewer: Include GitHub Username?
```suggestion
(This has been briefly discussed with Peter Beverloo (@beverloo), an editor of the Push API
```

Author: Done.

parameters to the `Badge.set` API. Upon receipt of the push, the user agent
would automatically call the `Badge` API without running any user code, and with
no requirement to show a notification.

This solution still has the following two flaws (expressed by Peter Beverloo):

1. The server can still use badge delivery to know whether device is online
   through [delivery receipts](https://tools.ietf.org/html/rfc8030#section-6.3).
   This may or may not be a privacy concern.
2. (On mobile) (at least on Android and iOS) the browser needs to be woken up to
   decrypt and process the payload; it can't be processed by the host OS's push
   receipt system due to the encryption. This could be a big resource drain if
   badges are pushed too frequently (the same concern applies for notifications,
   but as above, there is a natural tendency for developers to reduce the number
   of user-visible notifications due to user spam; the same forcing function
   does not apply for setting a badge).

#### Conclusion

Both of the above changes present huge obstacles and discussions involving
privacy, resource usage, and utility trade-offs. Due to the increased

> complexity, we are not considering changes to the Push API at this time.

Reviewer: This makes me sad and happy at the same time.

## Feature detection

Ideally we would not provide any feedback to the site as to how the badge is
going to be displayed to the user, since it's a user interface concern that
should be at the discretion of the user agent. For example, the site shouldn't
need to know whether the badge is displayed on a tab icon or on an app icon.

However, a practical concern overrides this: because the favicon can be (and
often is) used to display a page badge, sites upgrading to the `Badge` API need
to know whether they need to fall back to setting the favicon (since,
presumably, they don't want the badge being displayed twice). So we need to
provide a way to feature detect — not just "whether the Badge API is supported",
but "whether a Badge-API badge will show up on or near the favicon". So we

> introduce the `Badge.canBadgeDocument` API:

Reviewer: What about having `Badge.set()` return a list of places that were badged? 

or `Badge.setDocument(4)` returns a boolean success status?

Author: That's probably better.

Reviewer: Discussed doing this separately offline.

```js
if (Badge.canBadgeDocument()) {
  Badge.set(getUnreadCount(), {scopeDocument: true});
} else {
  // Fall back to favicon.
  showBadgeOnFavicon(getUnreadCount());
}
```

## A case for separation

Having enumerated all of the complexity relating to "document" versus "handle"

> contexts, I think we can make a case for separating them into two distinct APIs.

Reviewer: 👍

Under this alternative, there would be an API for setting a badge on URL scopes
that would *only* be displayed on URL handles such as apps and bookmarks, and
another API for setting a badge on the current document.

This would mean we no longer have a quick solution for setting the badge
everywhere on the current origin: to do that, you would need to call both APIs,
the latter on each page load.

The reason to do this is that the two APIs have very different concerns which
would be properly separated:

* The URL scoping complexity only applies to the "handle" API.
* The push messaging complexity only applies to the "handle" API.
* We wouldn't need `canBadgeDocument`; the presence of the "document" API itself

>   would indicate whether the tab will be badged in the UI.

Reviewer: Should we make a note that sites that do wish to do both could use some kind of third party library to do this?

The library could even polyfill the document badging, if it isn't available.

If we decide what our API should look like, I'm happy to throw one together.

Author: I guess the library would look like our old API and store its state in indexeddb or something so it can retrieve it on load and set it dynamically? Seems messy, since the indexeddb state could get out of sync with the user agent's state.

I think if we split the APIs, the best practices will be for developers to just treat them separately.

Other reasons to separate:

* Sites that only want to badge their app icon or only want to badge their tab
  could just concern themselves with the relevant API.
* The "document" API could potentially support badges that aren't supported by
  host operating systems, such as arbitrary Unicode characters.

## Detailed API proposal

### The model

> A badge is associated with a [scope](https://www.w3.org/TR/appmanifest/#navigation-scope).

Reviewer: ```suggestion
 A badge is associated with a [scope](https://w3c.github.io/ServiceWorker/#service-worker-registration-scope).
```

> * Documents are badged with the most specific badge for their URL (i.e. prefer a badge for `/page/1` to a badge for `/page/` when on the url `/page/1?foo=7`).

Reviewer: This is already discussed in the [Nested Scopes](#nested-scopes) section.

Author: That's OK. It needs to be summarized in the model. In some sense, this entire section is just a summary of what's been discussed above, showing the whole API surface in one place.

> * For [installed applications](https://www.w3.org/TR/appmanifest/#installable-web-applications), a user agent **MAY** display the badge with the most specific scope still encompassing the [navigation scope](https://www.w3.org/TR/appmanifest/#navigation-scope) of the application in an [OS specific context](#OS-Specific-Contexts).

Reviewer: I think this is already covered in the [Badge scope](#badge-scope) section. I don't know whether we need this bullet anymore.

Author: As above.

At any time, the badge for a specific scope, if it is set, may be either:

* A "flag" indicating the presence of a badge with no contents, or
* A positive integer.

The model does not allow a badge to be a negative integer, or the integer value 0 (setting the badge to 0 is equivalent to clearing the badge).

### The API

The `Badge` interface is a member object on
[`Window`](https://html.spec.whatwg.org/#the-window-object) and
[`ServiceWorkerGlobalScope`](https://w3c.github.io/ServiceWorker/#serviceworkerglobalscope-interface).
It contains the following methods:

* `Badge.set([contents], [options])`: Sets the badge for the scope in *options*
  to *contents* (an integer), or to "flag" if *contents* is omitted. If
  *contents* is 0, clears the badge for the given scope.
* `Badge.clear([options])`: Clears the badge for the scope in *options*.

> * `Badge.canBadgeDocument()`: Returns true if the user agent would display a

Reviewer: ```suggestion
* `Badge.canBadgeDocument()`: Returns true if the user agent can display a
    document scoped badge on or near the page's favicon.
```

Author: I think would is better here. If it merely can, then you have no way of knowing whether it did.

I guess your API is better where it just returns what it *did* do. But since this is a strawman proposal, I'll leave canBadgeDocument for now since I don't want to design a second strawman proposal for feature detection.

>   badge applying to a document scope on or near the page's favicon.

Reviewer: ```suggestion
```

The *options* parameter is a dictionary containing a single member, `scope`,
which contains a URL prefix to scope the badge to. If omitted, it defaults to
`"/"`.

> Note: Should we have a separate overload for boolean flags now, as discussed in [Issue 19](https://github.com/WICG/badging/issues/19) and [Issue 42](https://github.com/WICG/badging/issues/42)?

### UX treatment
Badges may appear in any place that the user agent deems appropriate. In general, these places should be obviously related to the pages being badged, so users understand what the status is for. Appropriate places could include:
- Tab favicons.
- Bookmark icons.

> - [OS Specific Contexts](#OS-Specific-Contexts) for [Installed Web Applications](https://www.w3.org/TR/appmanifest/#installable-web-applications).

Reviewer: Might be worth adding your NTP example here.

```suggestion
- Top sites, on the new tab page.
- [OS Specific Contexts](#OS-Specific-Contexts) for [Installed Web Applications](https://www.w3.org/TR/appmanifest/#installable-web-applications).
````

Author: Done.

> > Note: When showing an badge in an [OS Specific Context](#OS-Specific-Contexts) user agents should attempt reuse existing [operating system APIs and conventions](#Specific-operating-system-treatment-for-installed-web-applications), to achieve a native look-and-feel.

Reviewer: Link looks broken to me. Should maybe be to `docs/implementation.md`

Author: Done.

Limiting support to integers makes behavior more predictable, though we are considering
whether it might be worth adding support for other characters or symbols in future.

### Couldn’t this be a declarative API (i.e., a DOM element), so it would work without JavaScript?

> It could be, yes. However, as badges may be shared across multiple documents, this could be kind of confusing (e.g. there is a `<link rel="shortcut icon badge" href="/favicon.ico" badge="99">` in the head of a document, but it is being badged with 7 because another page was loaded afterwards). There is some discussion of this [here](https://github.com/WICG/badging/issues/1#issuecomment-485635068).

Reviewer: Is it worth reconsidering this for `document` badging, if we split them apart. I'm quite a fan.

Author: OK.
